### PR TITLE
20240104-with-liboqs-clang-tidy

### LINKED ISSUE #7108: [Bug]: liboqs build fails with clang
### Contact Details

_No response_

### Version

current master (commit 9db20774d8abe0725734d48090bd4f650a477b8a)

### Description

Building wolfssl with liboqs fails with clang (I'm using Clang 15) due to an implicit casting in `wolfcrypt/src/port/liboqs/liboqs.c`. Possibly the consequence of https://github.com/wolfSSL/wolfssl/pull/7026

### Reproduction steps

export CC=clang
git clone https://github.com/open-quantum-safe/liboqs.git
git clone --depth 1 https://github.com/wolfSSL/wolfssl.git
mkdir liboqs-install/
cd liboqs/
git checkout 0.8.0
mkdir build/
cd build/
cmake -DCMAKE_INSTALL_PREFIX:PATH=$(realpath ../../liboqs-install) -DOQS_USE_OPENSSL=0 ..
make -j$(nproc)
make install
cd ../../wolfssl/
autoreconf -ivf
./configure --enable-static --disable-examples --disable-crypttests --enable-aesccm --enable-aescfb --enable-aesctr --enable-aeseax --enable-aesgcm-stream --enable-aeskeywrap --enable-aesofb --enable-aessiv --enable-arc4 --enable-asn=original --enable-blake2 --enable-blake2s --enable-camellia --enable-certext --enable-cmac --enable-compkey --enable-crl --enable-cryptocb --enable-curve25519 --enable-curve448 --enable-des3 --enable-dsa --enable-dtls --enable-dtls13 --enable-dtlscid --enable-earlydata --enable-ecccustcurves --enable-ecccustcurves=all --enable-eccencrypt --enable-eccsi --enable-ed25519 --enable-ed25519-stream --enable-ed448 --enable-ed448-stream --enable-harden --enable-hkdf --enable-hrrcookie --enable-indef --enable-keygen --enable-md2 --enable-md4 --enable-nullcipher --enable-ocsp --enable-ocspstapling --enable-oldtls --enable-postauth --enable-psk --enable-pwdbased --enable-ripemd --enable-scrypt --enable-secure-renegotiation --enable-session-ticket --enable-shake128 --enable-shake256 --enable-siphash --enable-smallstack --enable-sni --enable-srp --enable-srtp --enable-sslv3 --enable-tls13 --enable-tlsv10 --enable-tlsx --enable-x963kdf --enable-xchacha --enable-xts --with-eccminsz=0 --enable-hpke --enable-ocspstapling2 --enable-pkcs7 --enable-dtls-frag-ch --with-liboqs=$(realpath ../liboqs-install/)
make

### Relevant log output

```shell
wolfcrypt/src/port/liboqs/liboqs.c:53:62: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'word32' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    int ret = wc_RNG_GenerateBlock(liboqsCurrentRNG, buffer, numOfBytes);
              ~~~~~~~~~~~~~~~~~~~~                           ^~~~~~~~~~
1 error generated.
```

### PR DESCRIPTION
**Pull request by douzzer**
fixes for `clang-tidy` reported defects and misstylings `--with-liboqs`:
* `readability-named-parameter` (style)
* `bugprone-sizeof-expression` (true bugs)
* `clang-analyzer-deadcode.DeadStores` (true bugs)
* `clang-analyzer-core.NonNullParamChecker` (true bug)
* `clang-diagnostic-newline-eof` (style)
* `clang-diagnostic-shorten-64-to-32` (true but benign in practice)

fixes for sanitizer reported defects `--with-liboqs`: null pointer `memcpy()`s in `TLSX_KeyShare_GenPqcKey()` and `server_generate_pqc_ciphertext()`.

fixes for silent crypto-critical failure in `wolfSSL_liboqsGetRandomData()`: refactor to accommodate oversize `numOfBytes`, and `abort()` if `wc_RNG_GenerateBlock()` returns failure.

tested with `wolfssl-multi-test.sh ... check-self check-file-modes check-source-text check-shell-scripts check-configure all-gcc-c99 pq-all pq-all-clang-tidy pq-all-sanitizer pq-all-cppcheck`

(`pq-all-clang-tidy` `pq-all-sanitizer` `pq-all-cppcheck` are new tests)

supersedes fix in #7111 which itself fixed #7108 .

### CODE REVIEW

**Comment by douzzer** [comment]
retest this please.

**Review by dgarske — APPROVED** [review]

**Comment by douzzer** [comment]
retest this please.

File: wolfcrypt/src/port/liboqs/liboqs.c [hunk 1]
```diff
@@ -50,9 +51,24 @@ static int liboqs_init = 0;
...
+    int ret;
+    word32 numOfBytes_word32;
+
+    while (numOfBytes > 0) {
+        numOfBytes_word32 = (word32)numOfBytes;
+        numOfBytes -= numOfBytes_word32;
+        ret = wc_RNG_GenerateBlock(liboqsCurrentRNG, buffer,
+                                   numOfBytes_word32);
+        if (ret != 0) {
+            /* ToDo: liboqs exits programm if RNG fails,
```
Comment (SparkiDev) [inline wolfcrypt/src/port/liboqs/liboqs.c:63]: You can do a limited number of retries, say 10.
If still fails, return failure and allow liboqs to exit.
We can't stop there bad behaviour but it is kinda fatal at this point.
Customers can fix liboqs if they don't like it.

**Review by SparkiDev — DISMISSED** [review]

File: wolfcrypt/src/port/liboqs/liboqs.c (hunk 1 above)
Comment (douzzer) [reply wolfcrypt/src/port/liboqs/liboqs.c:63]: Can't "return failure" -- liboqs in its infinite wisdom made the callback a `void`.  So it's either silently fail, or `abort()`, or something really complicated using a thread-local variable, which seems way out of scope for our purposes here.

Should we really bother retrying `wc_RNG_GenerateBlock()`?  Particularly without a `usleep` before each retry?  I guess you're thinking of `DRBG_NEED_RESEED`, but `wc_GenerateSeed()` uses `/dev/urandom` in the default and `WOLFSSL_GETRANDOM` implementations, so if they fail once they'll fail forever.

File: wolfcrypt/src/sphincs.c [hunk 2]
```diff
@@ -857,7 +857,7 @@ int wc_Sphincs_PrivateKeyDecode(const byte* input, word32* inOutIdx,
     else if ((key->level == 5) && (key->optim == FAST_VARIANT)) {
         keytype = SPHINCS_FAST_LEVEL5k;
     }
-    if ((key->level == 1) && (key->optim == SMALL_VARIANT)) {
+    else if ((key->level == 1) && (key->optim == SMALL_VARIANT)) {
```
Comment (anhu) [inline wolfcrypt/src/sphincs.c:860]: Nice catch!!

File: wolfcrypt/src/sphincs.c [hunk 3]
```diff
@@ -905,7 +905,7 @@ int wc_Sphincs_PublicKeyDecode(const byte* input, word32* inOutIdx,
...
```
Comment (anhu) [inline wolfcrypt/src/sphincs.c:908]: Nice catch!!

File: wolfcrypt/src/port/liboqs/liboqs.c (hunk 1 above)
Comment (anhu) [reply wolfcrypt/src/port/liboqs/liboqs.c:63]: I agree, not really worth doing anything complicated here.  However, can we remove the todo?  Nothing left to do.  It really is appropriate action.

**Review by anhu — CHANGES_REQUESTED** [review]

**Review by anhu — APPROVED** [review]

**Pull request merged**